crates/edr_eth/src/transaction/request/eip7702.rs

```diff
@@ -0,0 +1,176 @@
+use std::sync::OnceLock;
+
+use alloy_rlp::RlpEncodable;
+use revm_primitives::keccak256;
+
+use crate::{
+    eips::eip7702,
+    signature::{self, public_key_to_address, SecretKey, SignatureError, SignatureWithYParity},
+    transaction::{self, ComputeTransactionHash},
+    utils::envelop_bytes,
+    AccessListItem, Address, Bytes, B256, U256,
+};
+
+/// An [EIP-7702](https://eips.ethereum.org/EIPS/eip-7702) transaction.
+#[derive(Clone, Debug, Default, PartialEq, Eq, RlpEncodable)]
+pub struct Eip7702 {
+    // The order of these fields determines encoding order.
+    pub chain_id: u64,
+    pub nonce: u64,
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
+    pub gas_limit: u64,
+    pub to: Address,
+    pub value: U256,
+    pub input: Bytes,
+    pub access_list: Vec<AccessListItem>,
+    pub authorization_list: Vec<eip7702::SignedAuthorization>,
+}
+
+impl Eip7702 {
+    pub const TYPE: u8 = 4;
+
+    /// Signs the transaction with the provided secret key.
+    pub fn sign(
+        self,
+        secret_key: &SecretKey,
+    ) -> Result<transaction::signed::Eip7702, SignatureError> {
+        let caller = public_key_to_address(secret_key.public_key());
+
+        // SAFETY: The caller is derived from the secret key.
+        unsafe { self.sign_for_sender_unchecked(secret_key, caller) }
+    }
+
+    /// Signs the transaction with the provided secret key, belonging to the
+    /// provided caller's address.
+    ///
+    /// # Safety
+    ///
+    /// The `caller` and `secret_key` must correspond to the same account.
+    pub unsafe fn sign_for_sender_unchecked(
```

Comment:
Nit: this would be ideally restricted to `pub(super)` as it's dangereous, but I see that the other similar methods are public as well.

Reply:
It's marked as `unsafe` to indicate this danger. The safety doc explains what assumption exists about the relationship between the input arguments.

Reply:
Yeah I agree that this is ok if something has to be public, but wanted to point out that it's only used in the module, so it could be just as well `pub(super)`.